### PR TITLE
check for existing image before pushing to registry

### PR DESCRIPTION
**Pull request by pkwarren**
Update `buf alpha plugin push` to first check if an existing image digest exists in the OCI registry. If found, we can re-use it instead of pushing a new tag of the same image.

This should also greatly improve the performance of pushing images when the image already exists remotely.

### CODE REVIEW

File: private/buf/bufprint/curated_plugin_printer.go [hunk 1]
```diff
@@ -94,17 +94,19 @@ func (p *curatedPluginPrinter) printCuratedPluginsText(plugins ...*registryv1alp
...
 	return outputCuratedPlugin{
-		Owner:    plugin.Owner,
-		Name:     plugin.Name,
-		Version:  plugin.Version,
-		Revision: plugin.Revision,
+		Owner:       plugin.Owner,
+		Name:        plugin.Name,
+		Version:     plugin.Version,
+		Revision:    plugin.Revision,
+		ImageDigest: plugin.ContainerImageDigest,
```
Comment (pkwarren) [inline private/buf/bufprint/curated_plugin_printer.go:110]: Added the image digest to the JSON printed representation (useful for debugging).

File: private/buf/cmd/buf/command/alpha/plugin/pluginpush/pluginpush.go [hunk 2]
```diff
@@ -321,6 +295,121 @@ func run(
 	return bufprint.NewCuratedPluginPrinter(container.Stdout()).PrintCuratedPlugin(ctx, format, curatedPlugin)
 }
 
+func createCuratedPluginRequest(
```
Comment (pkwarren) [inline private/buf/cmd/buf/command/alpha/plugin/pluginpush/pluginpush.go:298]: Refactored this out to a helper method.

File: private/buf/cmd/buf/command/alpha/plugin/pluginpush/pluginpush.go [hunk 3]
```diff
@@ -321,6 +295,121 @@ func run(
...
+		Description:          plugin.Description(),
+		RegistryConfig:       protoRegistryConfig,
+		Revision:             nextRevision,
+		OutputLanguages:      outputLanguages,
+		SpdxLicenseId:        pluginConfig.SPDXLicenseID,
+		LicenseUrl:           pluginConfig.LicenseURL,
+	}, nil
+}
+
+func pushImage(
```
Comment (pkwarren) [inline private/buf/cmd/buf/command/alpha/plugin/pluginpush/pluginpush.go:328]: Refactored this out to a helper method.

File: private/buf/cmd/buf/command/alpha/plugin/pluginpush/pluginpush.go [hunk 4]
```diff
@@ -321,6 +295,121 @@ func run(
...
+// findExistingDigestForImageID will query the OCI registry to see if the imageID already exists.
+// If an image is found with the same imageID, its digest will be returned (and we'll skip pushing to OCI registry).
+//
+// It performs the following search:
+//
+// - GET /v2/{owner}/{plugin}/tags/list
+// - For each tag:
+//   - Fetch image: GET /v2/{owner}/{plugin}/manifests/{tag}
+//   - If image manifest matches imageID, we can use the image digest for the image.
+func findExistingDigestForImageID(ctx context.Context, plugin *bufpluginconfig.Config, imageID string) (string, error) {
```
Comment (pkwarren) [inline private/buf/cmd/buf/command/alpha/plugin/pluginpush/pluginpush.go:373]: This is the new method which looks up an existing digest for an image ID.

File: private/bufpkg/bufplugin/bufplugindocker/docker.go [hunk 5]
```diff
@@ -182,6 +190,14 @@ func (d *dockerAPIClient) Delete(ctx context.Context, image string) (*DeleteResp
 	return &DeleteResponse{}, nil
 }
 
+func (d *dockerAPIClient) Inspect(ctx context.Context, image string) (*InspectResponse, error) {
+	inspect, _, err := d.cli.ImageInspectWithRaw(ctx, image)
+	if err != nil {
+		return nil, err
+	}
+	return &InspectResponse{ImageID: inspect.ID}, nil
```
Comment (pkwarren) [inline private/bufpkg/bufplugin/bufplugindocker/docker.go:198]: Doing an inspect here allows us to check for an image ID in the OCI registry no matter if we pass a .zip file or an image with the `-image` flag.

**Review by mfridman — APPROVED** [review]

**Pull request merged**